postgis/lwgeom_in_marc21.c

```diff
@@ -0,0 +1,533 @@
+			min[len_min_sec] = '\0';
+
+			POSTGIS_DEBUGF(5, "  sec = palloc(%d)", literal_length);
+			sec = palloc(sizeof(char)*literal_length+1);
+			memcpy(sec, &literal[start_literal + 5],literal_length - (start_literal + 5));
+			sec[literal_length - (start_literal + 5)] = '\0';
+
+			result = atof(dgr) + (atof(min) / 60) + (atof(sec) / 3600);
+			POSTGIS_DEBUGF(2, "  minutes: %s", min);
+			POSTGIS_DEBUGF(2, "  decimal seconds: %s", sec);
+
+			POSTGIS_DEBUG(5, "  free(min)");
+			//pfree(min);
+			POSTGIS_DEBUG(5, "  free(dec)");
+			//pfree(sec);
+
+		}
+
+	}
+
+	/**
+	 * “+” for N and E (the plus sign is optional)
+	 * “-“ for S and W
+	 */
+
+	//pfree(dgr);
+
+	if (hemisphere_sign == 'S' || hemisphere_sign == 'W' || hemisphere_sign == '-') {
+
+		POSTGIS_DEBUGF(2, "  switching sign due to start character: '%c'", hemisphere_sign);
+		result = -result;
+
+	}
+
+	POSTGIS_DEBUGF(2, "=> parse_geo_literal returns: %.*f (in decimal degrees)", literal_length-(3+start_literal), result);
+	return result;
+}
+
+static LWGEOM*
+parse_marc21(xmlNodePtr xnode) {
+
+	int ngeoms;
+	int i;
+	xmlNodePtr datafield;
+	xmlNodePtr subfield;
+	LWGEOM *result;
+	LWGEOM **lwgeoms = (LWGEOM**) lwalloc(sizeof(LWGEOM*));
+	uint8_t geometry_type;
+	uint8_t result_type;
+	char *code;
+	char *literal;
+
+	POSTGIS_DEBUGF(2, "parse_marc21 called: root '<%s>'", xnode->name);
+
+	/**
+	 * MARC21/XML documents must have <record> as top level element.
+	 * https://www.loc.gov/standards/marcxml/xml/spy/spy.html
+	 */
+
+	if (xmlStrcmp(xnode->name, (xmlChar*) "record")) lwpgerror("invalid MARC21/XML document. Root element <record> expected but <%s> found.",xnode->name);
+
+	result_type = 0;
+	ngeoms = 0;
+
+	for (datafield = xnode->children; datafield != NULL; datafield = datafield->next) {
+
+		char *lw = NULL;
+		char *le = NULL;
+		char *ln = NULL;
+		char *ls = NULL;
+
+		if (datafield->type != XML_ELEMENT_NODE) continue;
+
+		if (xmlStrcmp(datafield->name, (xmlChar*) "datafield") != 0	|| xmlStrcmp(xmlGetProp(datafield, (xmlChar*) "tag"),(xmlChar*) "034") != 0) continue;
+
+		POSTGIS_DEBUG(3, "  datafield found");
+
+		for (subfield = datafield->children; subfield != NULL; subfield = subfield->next) {
+
+			if (subfield->type != XML_ELEMENT_NODE)	continue;
+			if (xmlStrcmp(subfield->name, (xmlChar*) "subfield") != 0) continue;
+
+			code = (char*) xmlGetProp(subfield, (xmlChar*) "code");
+
+			if ((strcmp(code, "d") != 0 && strcmp(code, "e") != 0 && strcmp(code, "f") != 0 && strcmp(code, "g")) != 0)	continue;
+
+			literal = (char*) xmlNodeGetContent(subfield);
+
+			POSTGIS_DEBUGF(3, "    subfield code '%s': %s", code, literal);
+
+			if (is_literal_valid(literal) == LW_TRUE) {
+
+				if (strcmp(code, "d") == 0) lw = literal;
+				else if (strcmp(code, "e") == 0) le = literal;
+				else if (strcmp(code, "f") == 0) ln = literal;
+				else if (strcmp(code, "g") == 0) ls = literal;
+
+			} else {
+
+				xmlFreeNode(subfield);
```

Comment:
don't you get this for free later when you free the whole document?

Reply:
Yes, you're right.. I'll remove the `xmlFreeNode(subfield);`